File: packages/tarka-ios/Sources/TarkaSDK/DeviceSignalCollector.swift
import CryptoKit
import Darwin
import Foundation
#if canImport(UIKit)
import UIKit
#endif

/// Collects iOS environment signals for [DeviceContext].
public final class DeviceSignalCollector {
    public init() {}

    public func collect() -> DeviceSignals {
        #if targetEnvironment(simulator)
        let isEmu = true
        #else
        let isEmu = false
        #endif

        let vpn = Self.isVpnActive()

        #if canImport(UIKit)
        let screen = "\(Int(UIScreen.main.bounds.width))x\(Int(UIScreen.main.bounds.height))"
        #else
        let screen: String? = nil
        #endif

        let lang = Locale.current.identifier

        return DeviceSignals(
            is_emulator: isEmu,
            is_vpn: vpn,
            is_spoofed_location: false,
            is_bot: false,
            is_repackaged: false,
            webdriver_detected: false,
            headless_detected: false,
            automation_detected: false,
            vpn_interface_detected: vpn,
            mock_location_detected: false,
            timezone_geo_mismatch: false,
            canvas_fp_hash: nil,
            webgl_renderer: nil,
            screen_res: screen,
            touch_support: true,
            battery_api_present: true,
            language: lang,
            platform_version: "iOS \(ProcessInfo.processInfo.operatingSystemVersionString)"
        )
    }

    public func computeDeviceId() -> String {
        #if canImport(UIKit)
        let id = UIDevice.current.identifierForVendor?.uuidString ?? "unknown"
        #else
        let id = "unknown"
        #endif
        let raw = "ios|\(id)"
        let digest = SHA256.hash(data: Data(raw.utf8))
        return digest.map { String(format: "%02x", $0) }.joined()
    }

    public func buildDeviceContext() -> DeviceContext {
        DeviceContext(device_id: computeDeviceId(), platform: "ios", signals: collect(), attestation: nil)
    }

    private static func isVpnActive() -> Bool {
        var addrs: UnsafeMutablePointer<ifaddrs>?
        guard getifaddrs(&addrs) == 0, let first = addrs else { return false }
        defer { freeifaddrs(first) }
        var ptr: UnsafeMutablePointer<ifaddrs>? = first
        while let p = ptr {
            let name = String(cString: p.pointee.ifa_name)
            if name.hasPrefix("utun") || name.hasPrefix("ppp") || name.hasPrefix("ipsec") {
                return true
            }
            ptr = p.pointee.ifa_next
        }
        return false
    }
}

File: packages/tarka-android/tarka-sdk/src/main/kotlin/io/tarka/sdk/DeviceSignalCollector.kt
package io.tarka.sdk

import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import android.provider.Settings
import java.security.MessageDigest

/**
 * Collects Android environment signals aligned with the Decision API [DeviceSignals] contract.
 * [deviceId] is a stable hash per install (not a hardware serial).
 */
class DeviceSignalCollector(private val context: Context) {

    fun collect(): DeviceSignals {
        val pm = context.packageManager
        val pkg = context.packageName
        val installer = try {
            @Suppress("DEPRECATION")
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                pm.getInstallSourceInfo(pkg).installingPackageName
            } else {
                pm.getInstallerPackageName(pkg)
            }
        } catch (_: Exception) {
            null
        }
        val expectedInstaller = "com.android.vending"
        val isRepackaged = installer != null && installer != expectedInstaller

        val isEmu = Build.FINGERPRINT.startsWith("generic")
            || Build.FINGERPRINT.startsWith("unknown")
            || Build.MODEL.contains("google_sdk")
            || Build.MODEL.contains("Emulator")
            || Build.MODEL.contains("Android SDK built for x86")
            || Build.MANUFACTURER.contains("Genymotion")
            || Build.BRAND.startsWith("generic") && Build.DEVICE.startsWith("generic")
            || "google_sdk" == Build.PRODUCT

        val devSettings = Settings.Global.getInt(
            context.contentResolver,
            Settings.Global.DEVELOPMENT_SETTINGS_ENABLED,
            0,
        ) == 1

        val vpn = try {
            val cm = context.getSystemService(Context.CONNECTIVITY_SERVICE) as? android.net.ConnectivityManager
            val active = cm?.activeNetwork
            val caps = if (active != null) cm.getNetworkCapabilities(active) else null
            caps?.hasTransport(android.net.NetworkCapabilities.TRANSPORT_VPN) == true
        } catch (_: Exception) {
            false
        }

        val mockLoc = try {
            Settings.Secure.getInt(context.contentResolver, "mock_location", 0) != 0
        } catch (_: Exception) {
            false
        }

        val lang = context.resources.configuration.locales[0]?.toString()

        return DeviceSignals(
            isEmulator = isEmu,
            isVpn = vpn,
            isSpoofedLocation = false,
            isBot = false,
            isRepackaged = isRepackaged,
            webdriverDetected = false,
            headlessDetected = false,
            automationDetected = devSettings,
            vpnInterfaceDetected = vpn,
            mockLocationDetected = mockLoc,
            timezoneGeoMismatch = false,
            canvasFpHash = null,
            webglRenderer = null,
            screenRes = screenResolution(),
            touchSupport = true,
            batteryApiPresent = true,
            language = lang,
            platformVersion = "Android ${Build.VERSION.RELEASE} (SDK ${Build.VERSION.SDK_INT})",
        )
    }

    private fun screenResolution(): String {
        val dm = context.resources.displayMetrics
        return "${dm.widthPixels}x${dm.heightPixels}"
    }

    /**
     * Deterministic device id: SHA-256 hex of package + installer + model + androidId (scoped).
     */
    fun computeDeviceId(): String {
        val androidId = Settings.Secure.getString(context.contentResolver, Settings.Secure.ANDROID_ID) ?: "unknown"
        val raw = listOf(
            context.packageName,
            Build.MODEL,
            Build.MANUFACTURER,
            androidId,
        ).joinToString("|")
        val md = MessageDigest.getInstance("SHA-256")
        val digest = md.digest(raw.toByteArray(Charsets.UTF_8))
        return digest.joinToString("") { "%02x".format(it) }
    }

    fun buildDeviceContext(): DeviceContext {
        val signals = collect()
        val id = computeDeviceId()
        return DeviceContext(deviceId = id, platform = "android", signals = signals, attestation = null)
    }
}
